paddle/fluid/operators/center_loss_op.h

```diff
@@ -0,0 +1,155 @@
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/operators/math/blas.h"
+#include "paddle/fluid/operators/math/functors.h"
+#include "paddle/fluid/platform/transform.h"
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenVector = framework::EigenVector<T, MajorType, IndexType>;
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
+
+template <typename T>
+struct SubFunctor {
+  inline HOSTDEVICE T operator()(T a, T b) const { return a - b; }
+};
+
+template <typename DeviceContext, typename T>
+class CenterLossKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &ctx) const override {
+    auto *X = ctx.Input<Tensor>("X");  // deep feature
+    auto *labels = ctx.Input<Tensor>("Label");
+    auto *centers = ctx.Input<Tensor>("Centers");
+    auto *update_rate = ctx.Input<Tensor>("CenterUpdateRate");
+    int cluster_num = ctx.Attr<int>("cluster_num");
+    auto *lr_center = update_rate->data<T>();
+    T alpha = lr_center[0];
+    bool need_update = static_cast<T>(ctx.Attr<bool>("need_update"));
+
+    auto x_data = X->data<T>();
+    auto label_data = labels->data<int64_t>();
+
+    auto centers_dim = centers->dims();
+    auto centers_data = centers->data<T>();
+
+    auto x_dims = X->dims();
+    int batch_size = x_dims[0];
+    int deep_feat_dim = x_dims[1];
+
+    auto centers_diff = ctx.Output<Tensor>("SampleCenterDiff");
+    auto centers_diff_data = centers_diff->mutable_data<T>(ctx.GetPlace());
+    auto *out_loss = ctx.Output<Tensor>("Loss");
+
+    auto *centers_out = ctx.Output<Tensor>("CentersOut");
+    auto *centers_out_data = centers_out->mutable_data<T>(ctx.GetPlace());
+
+    if (centers_out_data != centers_data) {
+      int size = centers_out->numel() * sizeof(T);
+      memcpy(centers_out_data, centers_data, size);
+    }
+
+    std::vector<int> center_update_count(cluster_num, 1);
+    auto &dev_ctx = ctx.template device_context<DeviceContext>();
+
+    auto loss_data = out_loss->mutable_data<T>(ctx.GetPlace());
+
+    Tensor centers_diffacc;  // used to accumulate all diff
+    auto centers_diffacc_data =
+        centers_diffacc.mutable_data<T>(centers_dim, ctx.GetPlace());
+    int numel = centers_diffacc.numel();
+    std::memset(centers_diffacc_data, 0, sizeof(T) * numel);
+
+    auto blas = math::GetBlas<DeviceContext, T>(ctx);
+    int tLabel;
+
+    T *x_index;
+    T *center_index;
+    T *center_loss_diff_index;
+    T *acc_index;
+    platform::Transform<DeviceContext> trans;
+
+    for (int i = 0; i < batch_size; ++i) {
+      tLabel = label_data[i];
+      center_update_count[tLabel]++;
+      x_index = const_cast<T *>(x_data) + i * deep_feat_dim;  // xi index
```

Comment:
`const_cast` is unnecessary, because it seems that this op would not change `x_data`. Try to make the type of `x_index` to be `const T *` instead of `T*`.

Reply:
done